src/hangar/__init__.py

```diff
@@ -1,16 +1,26 @@
 __version__ = '0.5.0'
-__all__ = ('make_torch_dataset', 'make_tf_dataset', 'Repository')
+__all__ = ('Repository', 'make_numpy_dataset', 'make_torch_dataset', 'make_tensorflow_dataset')
 
 from .repository import Repository
 
 
-def __getattr__(name):  # lazy loader
-    if name == 'make_tf_dataset':
-        from .dataloaders import make_tf_dataset
-        return make_tf_dataset
+def __dir__():
```

Comment:
@rlizzo For the reference, this makes hangar incompatible for python 3.6, AFIK

Reply:
We might have to go back to our old way of lazy loading with classes. What do you think?

Reply:
Hey so now that 0.5 is finally released I'm giving this the time it deserves.

Unfortunately I had to totally revert these changes since I didn't see a way around the py3.6 issues without messing with the `sys.path` config (a bit no-go for me), and it was even breaking 3.7/3.8 on windows. We've got a fairly efficient solution to a rather difficult problem; it does the job, and is easily understandable, this isn't an area of the codebase that needs to get complicated IMO. 

Style points aside, I do think we're generally OK with the status-quo from a usability perspective? The `make_foo_dataset` methods are exposed explicitly via the `__all__` variable.  They're available seamlessly in a script, and are visible / introspectable when working in the `repl`. Have you ran into issues anywhere?

Reply:
`__getattr__` & `__dir__` are introduced in [python 3.7](https://www.python.org/dev/peps/pep-0562/)

```python
from hangar import make_numpy_dataset  # will work
from hangar import make_torch_dataset  # won't work
from hangar import make_tensorflow_dataset  # won't work
```

This is my understanding. I haven't tried it out but I am pretty sure this is the case for python 3.6

Reply:
I actually wasn't aware of that PEP, but it seems that we've lucked out. 

When I was debugging it I actually saw the opposite

```python
from hangar import make_numpy_dataset  # won't work
from hangar import make_torch_dataset  # works
from hangar import make_tensorflow_dataset  # works 
```

Wether `make_numpy_dataset` was introduced in a guard didn't seem to matter (maybe since it would never raise an `ImportError`?). Nothing I did consistently worked across all the test platforms, so I dropped it. 

#### In any case, I think we lucked out in our current config.

> If an attribute is not found on a module object through the normal lookup (i.e. `object.__getattribute__`), then `__getattr__` is searched in the module `__dict__` before raising an `AttributeError`. If found, it is called with the attribute name and the result is returned. **Looking up a name as a module global will bypass module `__getattr__`. This is intentional, otherwise calling `__getattr__` for builtins will significantly harm performance.**

We set the module level global the first time that `__getattr__` is called. This makes alot more sense now about why that was necessary....

**If you run across another way to lazyload it might be work looking into**, I wasn't thrilled to read 

> This PEP may break code that uses module level (global) names `__getattr__` and `__dir__`. (**But the language reference explicitly reserves all undocumented dunder names, and allows "breakage without warning"; see [3].**)

Reply:
What's the deal with lazy loading anything? If the user tries to import a file that imports `torch` and they don't have torch installed, let them get an `ImportError`. If all this is about trying to massage the namespace so that everything either appears or doesn't in a top level module...  Don't do that.

It makes it harder to find the actual source code, and it confuses the issue, since if I'm missing another package, I'd like to get an ImportError that tells me that.

Reply:
#### Short answer:

- It's actually a major performance issue if we don't do some type of lazy-loading here. 

#### Longer Answer

In order to have the `make_torch_dataset` and `make_tf_dataset` methods available as a top-level import:

```python
from hangar import make_torch_dataset 
# compared to
from hangar.datasets.toch_loader import make_torch_dataset 
```
 we have to make them "importable" in the top-level `__init__.py` file. Since `torch` and `tensorflow` are optional dependencies, we have to at-least guard against environments missing one or both of the packages from raising `ImportError` / `ModuleNotFoundError` as hangar is imported. 

That in itself is fairly trivial to do. If the module exists then great, it gets imported as normal; if not, you can either mock it out, or use some `importlib`/`inspect` magic to cut the execution path short....

#### The issue for us isn't actually with environments that don't have `torch` or `tensorflow` installed. It's for the ones that do.*

These are really huge packages with a ton of dependencies themselves...  Importing them takes a noticeably long time, and having such massive dependencies sitting in RAM and polluting the package namespace accrues a non-trivial slowdown over time.

__**We use the "lazy-loaders" to stop the execution path of top-level hangar imports short of any file containing `import torch` or `import tensorflow` statments (unless explicitly requested). This occurs irregardless of if `torch` or `tensorflow` are installed in the environment.**__ The cost of keeping those packages around should only be paid if their actually being used :)

Reply:
@rlizzo I understand that you had reverted the guard with `__getattr__` but found that the functions are moved from the top level to inside `dataset`
```python
from hangar.dataset import make_xxx_dataset
```
Was that intentional. I'd still keep it on the top